samza-core/src/main/java/org/apache/samza/coordinator/JobCoordinatorMetadataManager.java

```diff
@@ -0,0 +1,283 @@
+    return metadata;
+  }
+
+  /**
+   * Persist the {@link JobCoordinatorMetadata} in metadata store. The job coordinator metadata is associated
+   * with the cluster type specified at the creation of the manager.
+   *
+   * @param metadata metadata to be persisted
+   *
+   * @return true if the write succeeded, false otherwise
+   */
+  public boolean writeJobCoordinatorMetadata(JobCoordinatorMetadata metadata) {
+    Preconditions.checkNotNull(metadata, "Job coordinator metadata cannot be null");
+
+    boolean writeSucceeded = false;
+    try {
+      String metadataValueString = metadataMapper.writeValueAsString(metadata);
+      metadataStore.put(clusterType, valueSerde.toBytes(metadataValueString));
+      writeSucceeded = true;
+    } catch (Exception e) {
+      LOG.error("Failed to write the job coordinator metadata to metadata store due to ", e);
+    }
+
+    LOG.info("Successfully written job coordinator metadata: {} for cluster {}.", metadata, clusterType);
+    return writeSucceeded;
+  }
+
+  @VisibleForTesting
+  Counter getApplicationAttemptCount() {
+    return applicationAttemptCount;
+  }
+
+  @VisibleForTesting
+  Gauge<Integer> getJobModelChangedAcrossApplicationAttempt() {
+    return jobModelChangedAcrossApplicationAttempt;
+  }
+
+  @VisibleForTesting
+  Gauge<Integer> getConfigChangedAcrossApplicationAttempt() {
+    return configChangedAcrossApplicationAttempt;
+  }
+
+  @VisibleForTesting
+  Gauge<Integer> getNewDeployment() {
+    return newDeployment;
+  }
+
+  @VisibleForTesting
+  String getEnvProperty(String propertyName) {
+    return System.getenv(propertyName);
+  }
+
+  /**
+   * Generate the epoch id using the execution container id that is passed through system environment. This isn't ideal
+   * way of generating this ID and we will need some contract between the underlying cluster manager and samza engine
+   * around what the epoch ID should be like and what is needed to generate is across different cluster offerings.
+   * Due to unknowns defined above, we leave it as is and keep it simple for now. It is favorable to keep it this way
+   * instead of introducing a loosely defined interface/API and marking it unstable.
+   *
+   * The properties of the epoch identifier are as follows
+   *  1. Unique across applications in the cluster
+   *  2. Remains unchanged within a single deployment lifecycle
+   *  3. Remains unchanged across application attempt within a single deployment lifecycle
+   *  4. Changes across deployment lifecycle
+   *
+   *  Note: The above properties is something we want keep intact when extracting this into a well defined interface
+   *  or contract for YARN AM HA to work.
+   *
+   * @return an identifier associated with the job coordinator satisfying the above properties
+   */
+  private String fetchEpochIdForJobCoordinator() {
+    String[] containerIdParts = getEnvProperty(CONTAINER_ID_PROPERTY).split(CONTAINER_ID_DELIMITER);
```

Comment:
just clarifying - this is very yarn specific right. should we call out this is ONLY for YARN at the moment?
and even within Yarn - it is very specific to the way yarn names containers/app attempts right now -- i dont forsee yarn changing this due to backwards compatibility but wanted to sure we are aware of this.